Add introductory text for models in documentation (#2416)
nest/nest-simulator: doc/htmldoc/models/models-main.rst

NESTML combines:

* fast prototyping
* an easy to understand, yet powerful syntax;
* a flexible processing toolchain, written in Python;
* good simulation performance by means of code generation (C++ for NEST Simulator).

.. seealso::

  :ref:`Install NESTML with NEST <nestml>`

Model naming
------------

Neuron models
~~~~~~~~~~~~~

Neuron model names in NEST combine abbreviations that describe the dynamics and synapse specifications for that model.
They may also include author's name of a model based on a specific paper.

Example neuron model name

``iaf_cond_beta``

    an implementation of a spiking neuron using integrate and fire dynamics with
    conductance-based synapses. Incoming spike events induce a postsynaptic change
    of conductance modelled by a beta function.

Example neuron model name based on specific paper

``hh_cond_exp_traub``

    a modified version of the Hodgkin Huxley neuron model based on Traub and Miles (1991)

Synapse models
~~~~~~~~~~~~~~

Synapse models include the word synapse as last word in the model name.

Synapse models may begin with author name (e.g., ``clopath_synapsae``) or process (e.g., ``stdp_synase``).

Devices
~~~~~~~

> A device name should represent its physical counter point - like a multimeter is ``multimeter``.  In general, the term ``recorder`` is used for devices

Author: d'oh, yes I'll fix that

> that store or generate output. The term  ``generator`` is used for devices that provide input into the simulation.

Author: Couldnt the output be various forms of data? would providing a couple of examples work?

> that store or generate output. The term  ``generator`` is used for devices that provide input into the simulation.

Reviewer: sure, perhaps something like "output such as spike times or synaptic strengths over time"?

> that store or generate output. The term  ``generator`` is used for devices that provide input into the simulation.

Author: done

Table of model terms
~~~~~~~~~~~~~~~~~~~~~

The following table provides the NEST term used for a specific descriptor of a model along with commonly used alternative terms you may find in the literature.
Additional terms may be used to help differentiate models not included in the list below.
See the the :doc:`model directory <index>` to access the full description of each model.

> 

Reviewer: link to glossary instead/additionally?

Author: yes, I think instead of table here it would better to put terms in a subsection of glossary and link to that.

Author: @clinssen I reorganized glossary into subsections and linked to the a section on *terms for models*

.. _modelsmain:

Models in NEST
==============

What we mean by `models`
------------------------

> Models in the context of NEST are C++ implmentations of mathematical equations that describe the characteristics and behavior of

Reviewer 2: ```suggestion
Models in the context of NEST are C++ implementations of mathematical equations that describe the characteristics and behavior of
```

> different types of neurons, syanpses, (and devices?), based on the relevant peer-reviewed publications for the model.

Reviewer 2: ```suggestion
different types of neurons, synapses, (and devices?), based on the relevant peer-reviewed publications for the model.
```

We also use the term model in relation to network models (e.g., :ref:`microcircuit <toc_microcircuit>` and `multi-area model <https://inm-6.github.io/multi-area-model/>`_). These network models

> can be considerd a level of complexity higher than the neuron or synapse model. We do not focus on network models here.

Reviewer 2: ```suggestion
can be considered a level of complexity higher than the neuron or synapse model. We do not focus on network models here.
```

Find a model
------------

NEST provides a ton of models! Textbook standards like integrate-and-fire and Hodgkin-Huxley type models are available
alongside high quality implementations of models published by the neuroscience community.

> The models are organized and autogenerated by keywords (e.g., adaptive threshold, conductance-based etc) in our directory. Models that contan a specific keyword will be listed under that word.

Reviewer: can keyword examples straightaway be links?

Author: yes

> The models are organized and autogenerated by keywords (e.g., adaptive threshold, conductance-based etc) in our directory. Models that contan a specific keyword will be listed under that word.

Reviewer 2: ```suggestion
The models are organized and autogenerated by keywords (e.g., adaptive threshold, conductance-based etc) in our directory. Models that contain a specific keyword will be listed under that word.
```

* Discover :doc:`all the models in our directory <index>`.

> NESTML

Reviewer: Maybe this paragraph as an admonition (``.. hint``)?

Author: done

------

> * Check out :doc:`NESTML <nestml:index>`,  a domain-specific language for neuron and synapse models. These dynamical models can

Reviewer: ```suggestion
* Check out :doc:`NESTML <nestml:index>`, a domain-specific language for neuron and synapse models. NESTML enables fast prototyping of new models using an easy to understand, yet powerful syntax, and combines this with high simulation performance through the automated generation of C++ code, suitable for use in NEST Simulator.
```

  be used in simulations of brain activity on several platforms, in particular the NEST Simulator.

NESTML combines:

* fast prototyping
* an easy to understand, yet powerful syntax;
* a flexible processing toolchain, written in Python;

> * good simulation performance by means of code generation (C++ for NEST Simulator).

Reviewer 2: Can we also give some instructions on how to install NESTML here or on the install page?

Author: done

Model naming
------------

Neuron models
~~~~~~~~~~~~~

Neuron model names in NEST combine abbreviations that describe the dynamics and synapse specifications for that model.
They may also include author's name of a model based on a specific paper.

Example neuron model name

``iaf_cond_beta``

    an implementation of a spiking neuron using integrate and fire dynamics with
    conductance-based synapses. Incoming spike events induce a postsynaptic change
    of conductance modelled by a beta function.

Example neuron model name based on specific paper

``hh_cond_exp_traub``

    a modified version of the Hodgkin Huxley neuron model based on Traub and Miles (1991)

Synapse models
~~~~~~~~~~~~~~

> Synapse models  include the word synapse as last word in model name.

Reviewer 2: ```suggestion
Synapse models include the word synapse as last word in the model name.
```

.. _modelsmain:

Models in NEST
==============

What we mean by `models`
------------------------

Models in the context of NEST are C++ implementations of mathematical equations that describe the characteristics and behavior of
different types of neurons and synapses, based on the relevant peer-reviewed publications for the model.

We also use the term model in relation to network models (e.g., :ref:`microcircuit <toc_microcircuit>` and `multi-area model <https://inm-6.github.io/multi-area-model/>`_). These network models
can be considered a level of complexity higher than the neuron or synapse model. We do not focus on network models here.

Find a model
------------

NEST provides a ton of models! Textbook standards like integrate-and-fire and Hodgkin-Huxley type models are available
alongside high quality implementations of models published by the neuroscience community.

> The models are organized and autogenerated by keywords (e.g., :doc:`adaptive threshold <index_adaptive threshold>`, :doc:`conductance-based <index_conductance-based>` etc) in our directory. 

Author: fixed

Models that contain a specific keyword will be listed under that word.

.. seealso::

   Discover :doc:`all the models in our directory <index>`.

Create and customize models with NESTML
---------------------------------------

> Check out :doc:`NESTML <nestml:index>`, a domain-specific language for neuron and synapse models.

Reviewer 3: What about this comment?

NESTML combines:

* fast prototyping
* an easy to understand, yet powerful syntax;
* a flexible processing toolchain, written in Python;
* good simulation performance by means of code generation (C++ for NEST Simulator).

.. seealso::

  See the :doc:`NESTML docs for installation details <nestml:index>`

.. note::

  NESTML is also available as part of NEST's official :ref:`docker image <docker>`

Model naming
------------

Neuron models
~~~~~~~~~~~~~

Neuron model names in NEST combine abbreviations that describe the dynamics and synapse specifications for that model.
They may also include author's name of a model based on a specific paper.

Example neuron model name

``iaf_cond_beta``

    an implementation of a spiking neuron using integrate and fire dynamics with
    conductance-based synapses. Incoming spike events induce a postsynaptic change
    of conductance modelled by a beta function.

Example neuron model name based on specific paper

``hh_cond_exp_traub``

    a modified version of the Hodgkin Huxley neuron model based on Traub and Miles (1991)

Synapse models
~~~~~~~~~~~~~~

Synapse models include the word synapse as last word in the model name.

> Synapse models may begin with author name (e.g., ``clopath_synapsae``) or process (e.g., ``stdp_synase``).

Reviewer 2: ```suggestion
Synapse models may begin with author name (e.g., ``clopath_synapse``) or process (e.g., ``stdp_synapse``).
```

Devices
~~~~~~~

A device name should represent its physical counterpart - like a multimeter is ``multimeter``.  In general, the term ``recorder`` is used for devices

> that store or generate output (e.g., spike times or synaptic strengths over time). The term  ``generator`` is used for devices that provide input into the simulation.

Reviewer 3: ```suggestion
that store the output (e.g., spike times or synaptic strengths over time) of other nodes and make it accessible to the user. The term  ``generator`` is used for devices that provide input into the simulation.
```

Reviewer 3: Do the recorders really generate output? If yes, please ignore my suggestion.

.. seealso::

>   See our glossary section on :ref:`common abbreviations used for model terms <model_terms>`. It includes alternative terms commonly used in the literatue.

Reviewer 2: ```suggestion
  See our glossary section on :ref:`common abbreviations used for model terms <model_terms>`. It includes alternative terms commonly used in the literature.
```

NESTML enables fast prototyping of new models using an easy to understand, yet powerful syntax, and combines

> this with high simulation performance through the automated generation of C++ code, suitable for use in NEST Simulator.

Reviewer 3: Third point got lost.

> this with high simulation performance through the automated generation of C++ code, suitable for use in NEST Simulator.

Reviewer 3: ```suggestion
NESTML enables fast prototyping of new models using an easy to understand, yet powerful syntax. This is achieved by a combination of a flexible processing toolchain
written in Python with high simulation performance through the automated generation of C++ code, suitable for use in NEST Simulator.
```

Reviewer 3: The third bulletpoint got lost. Tried to add it back in.
